File: tests/abi/plugview_layout.cpp
#include "pluginterfaces/gui/iplugview.h"
#include "pluginterfaces/gui/iplugviewcontentscalesupport.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(Type) std::printf (#Type " size %zu align %zu\n", sizeof (Steinberg::Type), alignof (Steinberg::Type))
#define PRINT_OFFSET(Type, Field) std::printf (#Type "." #Field " offset %zu\n", offsetof (Steinberg::Type, Field))

template <typename T>
static void print_iid (const char* name, const T& tuid)
{
	std::printf ("%s iid", name);
	for (int index = 0; index < 16; ++index)
		std::printf (" %02X", static_cast<unsigned char> (tuid[index]));
	std::printf ("\n");
}

int main ()
{
	std::printf ("Platform.kPlatformTypeHWND %s\n", Steinberg::kPlatformTypeHWND);
	std::printf ("Platform.kPlatformTypeHIView %s\n", Steinberg::kPlatformTypeHIView);
	std::printf ("Platform.kPlatformTypeNSView %s\n", Steinberg::kPlatformTypeNSView);
	std::printf ("Platform.kPlatformTypeUIView %s\n", Steinberg::kPlatformTypeUIView);
	std::printf ("Platform.kPlatformTypeX11EmbedWindowID %s\n", Steinberg::kPlatformTypeX11EmbedWindowID);
	std::printf ("Platform.kPlatformTypeWaylandSurfaceID %s\n", Steinberg::kPlatformTypeWaylandSurfaceID);

	PRINT_TYPE (ViewRect);
	PRINT_OFFSET (ViewRect, left);
	PRINT_OFFSET (ViewRect, top);
	PRINT_OFFSET (ViewRect, right);
	PRINT_OFFSET (ViewRect, bottom);

	PRINT_TYPE (IPlugView);
	PRINT_TYPE (IPlugFrame);
	PRINT_TYPE (Linux::IEventHandler);
	PRINT_TYPE (Linux::ITimerHandler);
	PRINT_TYPE (Linux::IRunLoop);
	PRINT_TYPE (IPlugViewContentScaleSupport);

	print_iid ("IPlugView", Steinberg::IPlugView_iid);
	print_iid ("IPlugFrame", Steinberg::IPlugFrame_iid);
	print_iid ("IEventHandler", Steinberg::Linux::IEventHandler_iid);
	print_iid ("ITimerHandler", Steinberg::Linux::ITimerHandler_iid);
	print_iid ("IRunLoop", Steinberg::Linux::IRunLoop_iid);
	print_iid ("IPlugViewContentScaleSupport", Steinberg::IPlugViewContentScaleSupport_iid);
	return 0;
}

File: tools/plugview_layout.zig
const std = @import("std");
const plugview = @import("zig-vst3").pluginterfaces.gui.iplugview;
const scale = @import("zig-vst3").pluginterfaces.gui.iplugviewcontentscalesupport;

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    try stdout.print("Platform.kPlatformTypeHWND {s}\n", .{std.mem.span(plugview.PlatformType.kPlatformTypeHWND)});
    try stdout.print("Platform.kPlatformTypeHIView {s}\n", .{std.mem.span(plugview.PlatformType.kPlatformTypeHIView)});
    try stdout.print("Platform.kPlatformTypeNSView {s}\n", .{std.mem.span(plugview.PlatformType.kPlatformTypeNSView)});
    try stdout.print("Platform.kPlatformTypeUIView {s}\n", .{std.mem.span(plugview.PlatformType.kPlatformTypeUIView)});
    try stdout.print("Platform.kPlatformTypeX11EmbedWindowID {s}\n", .{std.mem.span(plugview.PlatformType.kPlatformTypeX11EmbedWindowID)});
    try stdout.print("Platform.kPlatformTypeWaylandSurfaceID {s}\n", .{std.mem.span(plugview.PlatformType.kPlatformTypeWaylandSurfaceID)});

    try printType(stdout, "ViewRect", plugview.ViewRect);
    try printOffset(stdout, "ViewRect", "left", plugview.ViewRect, "left");
    try printOffset(stdout, "ViewRect", "top", plugview.ViewRect, "top");
    try printOffset(stdout, "ViewRect", "right", plugview.ViewRect, "right");
    try printOffset(stdout, "ViewRect", "bottom", plugview.ViewRect, "bottom");

    try printType(stdout, "IPlugView", plugview.IPlugView);
    try printType(stdout, "IPlugFrame", plugview.IPlugFrame);
    try printType(stdout, "Linux::IEventHandler", plugview.Linux.IEventHandler);
    try printType(stdout, "Linux::ITimerHandler", plugview.Linux.ITimerHandler);
    try printType(stdout, "Linux::IRunLoop", plugview.Linux.IRunLoop);
    try printType(stdout, "IPlugViewContentScaleSupport", scale.IPlugViewContentScaleSupport);

    try printTuid(stdout, "IPlugView", plugview.iplug_view_iid);
    try printTuid(stdout, "IPlugFrame", plugview.iplug_frame_iid);
    try printTuid(stdout, "IEventHandler", plugview.ievent_handler_iid);
    try printTuid(stdout, "ITimerHandler", plugview.itimer_handler_iid);
    try printTuid(stdout, "IRunLoop", plugview.irun_loop_iid);
    try printTuid(stdout, "IPlugViewContentScaleSupport", scale.iplug_view_content_scale_support_iid);
}

fn printType(writer: anytype, comptime name: []const u8, comptime Type: type) !void {
    try writer.print("{s} size {} align {}\n", .{ name, @sizeOf(Type), @alignOf(Type) });
}

fn printOffset(writer: anytype, comptime type_name: []const u8, comptime field_label: []const u8, comptime Type: type, comptime field_name: []const u8) !void {
    try writer.print("{s}.{s} offset {}\n", .{ type_name, field_label, @offsetOf(Type, field_name) });
}

fn printTuid(writer: anytype, comptime name: []const u8, bytes: [16]u8) !void {
    try writer.print("{s} iid", .{name});
    for (bytes) |byte| {
        try writer.print(" {X:0>2}", .{byte});
    }
    try writer.writeByte('\n');
}
